### PR TITLE
Fix registered wildcard clusterEvents doesn't work in scheduler requeueing

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?

/kind bug
/sig scheduling

#### What this PR does / why we need it:

As mentioned above, we ignored the wildcard events in scheduler, consider these two scenarios:
- registered a ClusterEvent with {Resource: "*", ActionType: All}
- registered a ClusterEvent with {Resource: "*', ActionType: Add}

Both don't match with a coming event {Resource: "Pod", ActionType: Add}, but actually they should.

/assign @sanposhiho 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug where registered wildcard clusterEvent sources didn't work in scheduler requeueing.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/types.go [hunk 1]
```diff
@@ -144,6 +144,13 @@ func (ce ClusterEvent) IsWildCard() bool {
 	return ce.Resource == WildCard && ce.ActionType == All
 }
 
+// Match returns true if ClusterEvent is matched with the coming event.
+// If the ce.Resource is "*", there's no requirement for the coming event' Resource.
+// Contrarily, if the coming event's Resource is "*", the ce.Resource should only be "*".
+func (ce ClusterEvent) Match(event ClusterEvent) bool {
+	return ce.IsWildCard() || (ce.Resource == WildCard || ce.Resource == event.Resource) && ce.ActionType&event.ActionType != 0
```
Comment (sanposhiho) [inline pkg/scheduler/framework/types.go:151]: Should we? (I know this suggestion wouldn't change anything in `isPodWorthRequeuing` though.)

```suggestion
	return ce.IsWildCard() || event.IsWildCard() || (ce.Resource == WildCard || ce.Resource == event.Resource) && ce.ActionType&event.ActionType != 0
```

File: pkg/scheduler/framework/types.go
Comment (sanposhiho) [inline pkg/scheduler/framework/types.go:1]: (GitHub doesn't allow me to comment at non-changed place)

Can we add a note for wildcard events at [here](https://github.com/kubernetes/kubernetes/blob/d81023db30cf91886706cc320cf23ca44c6539ca/pkg/scheduler/framework/types.go#L86)? I'd like to mention two usecases; when it's registered in EventsToRegister and when it's used by the scheduling queue - Something like:

```go
  // WildCard is a special GVK to match all resources' events.
  // e.g., If you register `{Resource: "*", ActionType: All}` in EventsToRegister, all resources' all events are subscribed by the entry.
  // Meaning, if you use it with QueueingHint, you have to consider that your QueueingHint would receive various objects.
  // 
  // Also, the scheduling queue also uses the wildcard event to move all Pods 
  // from the unschedulable pod pool to activeQ/backoffQ forcibly.  
  // In such cases, any QueueingHints are triggered so that all Pods are moved.
  WildCard              GVK = "*"
```

**Review by sanposhiho — COMMENTED** [review]
Thanks, it's a great catch.

File: pkg/scheduler/framework/types.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/types.go:151]: I haven't seen emitting a wildcard event yet, what does the coming event is a wildcard means? **We can definitely know what the coming event is**, so let's get rid of the " * " .
TL;DR
When the coming event is "*", it will only matched with `ce.Resource == event.Resource) && ce.ActionType&event.ActionType != 0`, make sense?

**Comment by sanposhiho** [comment]
Looking at https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-node-e2e-containerd, seems there's something wrong on-going with this job - all jobs are red. 😓

File: pkg/scheduler/framework/types.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/types.go:151]:
```
		// If the wildcard event is special one as someone wants to force all Pods to move to activeQ/backoffQ.
		// We return queueAfterBackoff in this case, while resetting all blocked plugins.
```

Got the idea here, but I still take it as an unmatched event but we have a special case here. I'll add the comment.

Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:151]: > I haven't seen emitting a wildcard event yet, what does the coming event is a wildcard means?

My comment at https://github.com/kubernetes/kubernetes/pull/123117#discussion_r1477207815 describes that -- the scheduling queue uses the wildcard event when moving Pods from unsched to backoffQ forcibly. ([`flushUnschedulablePodsLeftover`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L841))
Given this usecase, we've defined an incoming wildcard event to match all registered events and ignore QueueingHInts.

File: pkg/scheduler/framework/types.go
Comment (kerthcet) [reply pkg/scheduler/framework/types.go:1]: Make sense.

**Comment by kerthcet** [comment]
> Looking at https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-node-e2e-containerd, seems there's something wrong on-going with this job - all jobs are red. 😓

em.. I'll take a look.

**Comment by kerthcet** [comment]
Seems not related, any way I updated your comments, plz take a look @sanposhiho

File: pkg/scheduler/framework/types.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/types.go:163]: Yes, I didn't know that before, something new learned.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -441,7 +441,7 @@ func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework
 	pod := pInfo.Pod
 	queueStrategy := queueSkip
 	for eventToMatch, hintfns := range hintMap {
-		if eventToMatch.Resource != event.Resource || eventToMatch.ActionType&event.ActionType == 0 {
+		if !eventToMatch.Match(event) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:444]: Does this code run if the feature gate is disabled?

That would decide whether we need to cherry-pick to older versions.

Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: Alternatively, we don't need to cherry pick if we don't use these kind of wildcards in the default scheduler. Do we?

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: I don't think it's worth being cherry-picked.

We do reach here regardless of the feature gate though, no in-tree plugin uses this in `EventsToRegister`. Plus, I would say it was rare to use a wildcard event in `EventsToRegister` actually, since registering a wildcard equals leaving `EventsToRegister` empty. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework.go#L622-L626

Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: So, should we tag this like so?
````
/remove-kind bug
/kind cleanup
```

Comment (kerthcet) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: I take it as a non-critical bug as https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md#what-kind-of-prs-are-good-for-cherry-picks describes, make sense?

Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: SGTM

**Comment by sftim** [comment]
Change log suggestion
```diff
-Fix registered wildcard clusterEvents doesn't work in scheduler requeueing.
+Fixed a bug where registered wildcard cluster event sources didn't work in scheduler requeueing.
```

or

```diff
-Fix registered wildcard clusterEvents doesn't work in scheduler requeueing.
+Fixed a bug where registered `ClusterEvent` sources didn't work in scheduler requeueing
+if they used a wildcard match.
```

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:444]: +1
